ctru-rs/src/services/ir_user.rs

```diff
@@ -0,0 +1,496 @@
+                    return;
+                };
+
+                (move || unsafe {
+                    // Close service and memory handles
+                    ResultCode(ctru_sys::svcCloseHandle(shared_mem.service_handle))?;
+                    ResultCode(ctru_sys::svcCloseHandle(shared_mem.shared_memory_handle))?;
+
+                    // Free shared memory
+                    std::alloc::dealloc(
+                        shared_mem.shared_memory.as_ptr() as *mut u8,
+                        shared_mem.shared_memory_layout,
+                    );
+
+                    Ok(())
+                })()
+                .unwrap();
+            },
+        )?;
+
+        Ok(IrUser {
+            _service_reference: service_reference,
+        })
+    }
+
+    /// Try to connect to the device with the provided ID.
+    pub fn require_connection(&self, device_id: IrDeviceId) -> crate::Result<()> {
+        unsafe {
+            self.send_service_request(
+                vec![REQUIRE_CONNECTION_COMMAND_HEADER, device_id.get_id()],
+                2,
+            )?;
+        }
+        Ok(())
+    }
+
+    /// Close the current IR connection.
+    pub fn disconnect(&self) -> crate::Result<()> {
+        unsafe {
+            self.send_service_request(vec![DISCONNECT_COMMAND_HEADER], 2)?;
+        }
+        Ok(())
+    }
+
+    /// Get an event handle that activates on connection status changes.
+    pub fn get_connection_status_event(&self) -> crate::Result<Handle> {
+        let response = unsafe {
+            self.send_service_request(vec![GET_CONNECTION_STATUS_EVENT_COMMAND_HEADER], 4)
+        }?;
+        let status_event = response[3] as Handle;
+
+        Ok(status_event)
+    }
+
+    /// Get an event handle that activates when a packet is received.
+    pub fn get_recv_event(&self) -> crate::Result<Handle> {
+        let response =
+            unsafe { self.send_service_request(vec![GET_RECEIVE_EVENT_COMMAND_HEADER], 4) }?;
+        let recv_event = response[3] as Handle;
+
+        Ok(recv_event)
+    }
+
+    /// Circle Pad Pro specific request.
+    ///
+    /// This will send a packet to the CPP requesting it to send back packets
+    /// with the current device input values.
+    pub fn request_input_polling(&self, period_ms: u8) -> crate::Result<()> {
+        let ir_request: [u8; 3] = [1, period_ms, (period_ms + 2) << 2];
+        unsafe {
+            self.send_service_request(
+                vec![
+                    SEND_IR_NOP_COMMAND_HEADER,
+                    ir_request.len() as u32,
+                    2 + (ir_request.len() << 14) as u32,
+                    ir_request.as_ptr() as u32,
+                ],
+                2,
+            )?;
+        }
+
+        Ok(())
+    }
+
+    /// Mark the last `packet_count` packets as processed, so their memory in
+    /// the receive buffer can be reused.
+    pub fn release_received_data(&self, packet_count: u32) -> crate::Result<()> {
```

Comment:
Should this function require `&mut self`? Anything that changes internal states should require `mut`.